### ISSUE TITLE
Support @reboot command

### ISSUE THREAD

**Issue by hongkongkiwi** [aptible/supercronic#30]
Hi there,

I'm currently running supercronic in a docker and it's working great.

I would like to run my script on "boot" e.g. when supercronic is started, as well as at a set interval. Normally in many cron versions we can use the standard @reboot to run a script on reboot, but in my testing it looks like supercronic doesn't currently support this.

Would it be possible to add this functionality? I notice that other shortcuts such as @hourly are supported.

**Comment by krallin** [comment]
What's your expectation for what `@reboot` would do? Run the job upon Supercronic starting up?

(I'm a little concerned about this being potentially confusing, since that's not what you'd normally expect from a `@reboot` in regular cron)

**Comment by hongkongkiwi** [comment]
Yes that is right, I would expect supercronic to run the script when it first starts up.

That means I can have both an hourly and a startup command all in the same cron file (which makes it pretty clear).

**Comment by cgons** [comment]
I'm looking for this feature as well. 

Maybe rather than `@reboot` we call it `@cronstart` so as not to confuse with the act of a system reboot.

@krallin - Would you be open to a PR adding this feature?

**Comment by krallin** [comment]
@cgons yes, with the minor caveat that I think `@start` would be a simpler / appropriate name for the event. We'll want to make sure to document that this is non-standard though.

**Comment by hongkongkiwi** [comment]
Sounds like a great solution! Looking forward to the PR.

**Comment by cgons** [comment]
Awesome, `@start` it is!  I'll try and get started ASAP. 

BTW, is this the correct place to talk about implementation details (ie what I plan to change/add etc...)?

**Comment by krallin** [comment]
@cgons sure, that's fine :+1:

**Comment by hongkongkiwi** [comment]
Any progress on this @cgons ? I work around this by adding a docker-entrypoint.sh file, but I would love to get rid of this extra process handling and have everything done within supercronic.

**Comment by cgons** [comment]
Sorry, I had to go an pat. leave for a few months and ended up having negative free time to work on this.  I'm back now and I want to try and get this working over the next few days/weeks. 

I'll post here with an update. Be on my case though as that's a motivator to get this implemented! :)

**Comment by cgons** [comment]
@krallin - I was thinking of creating a `PredefinedExpression` class that implements the `Expression` interface and returns zero value time after it has been run once. 

Inside `parseJobLine()` I would check if the line starts with `@` and if so, attempt to parse it as a `PredefinedExpression`.   If the `@` does not match the list of predefined expressions (which for now would be just `@start`), then it would return a error and we can then attempt to parse it using `cronexpr`. 

I have to think a little more about how to keep track of when the job was run and where to store that state.  I should probably take advantage of the `cronIteration` counter...

Let me know what you think.

**Comment by krallin** [comment]
I think you should instead create a separate interface that wraps the existing cron expression interface, because you're also going to need a `IsDone` method or some such to represent that the `StartJob` goroutine should exit after running this job once.


I'm not clear on why you suggest `Next()` should return 0. I'm not certain I understand what this achieves here?

**Comment by romain-dartigues** [comment]
> yes, with the minor caveat that I think `@start` would be a simpler / appropriate name for the event. We'll want to make sure to document that this is non-standard though.

I'm not really fan of having a new (non-standard, `supercronic`-specific) directive, like `@start`.
`@reboot` is already documented as "**schedule this job only once, when crond starts up**" in [dcron](/dubiousjim/dcron).

The [Vixie cron manual](http://man7.org/linux/man-pages/man5/crontab.5.html) (which, AFAIK introduced this directive) just says "run once after reboot"; but the [debian version of the manual](https://manpages.debian.org/stretch/cron/crontab.5.en.html) page states: "Please note that startup, as far as `@reboot` is concerned, is the time when the **cron(8) daemon startup**. In particular, it may be before some system daemons, or other facilities, were startup. This is due to the boot order sequence of the machine.".

**Comment by dmaj** [comment]
Sorry, couldn't wait until someone started the implementation. I implemented the issue. Unfortunately this was my first contact with Golang. So please don't expect too much :-). Besides, it is somehow a hack ... But for my purposes it works.

I have extended the translation table in Cronexpr: "@start", "*/1 * * * * * * *".
So this job starts every second.
In the supercronic I check if the job was generated from an @start and terminate it after a single run.
To leave some time for @start processing, I introduced a parameter "-delay".
The line with @start must be at the first position.

https://github.com/dmaj/supercronic
https://github.com/dmaj/cronexpr

Translated with www.DeepL.com/Translator

**Comment by fbartels** [comment]
Maybe this is interesting for others as well. I solved the @reboot topic for me by simply doing a one off execution of my cron tasks before actually starting supercronic. This is easily doable since I am not mounting my crontab file into the container, but actually generate it at container startup. This has for me the added benefit that I am making sure that the cron jobs work at all (else the container would fail to start).

Code is at https://github.com/zokradonh/kopano-docker/blob/e0792a39a5d5ea83203647f303350101df97e590/scheduler/start.sh#L35

**Comment by phin1x** [comment]
i created a pull request to implement this issue. in the first run of the main loop i execute all jobs with the expression @start. after then the other periodic jobs will be scheduled.
the pr is reload safe.

**Comment by hongkongkiwi** [comment]
Awesome! Can we get this merged?

**Comment by MikeC-BC** [comment]
would love to see this merged

**Comment by dex4er** [comment]
Actually I would prefer `@reboot` instead of `@start` as it is supported by all cron replacements. You can just make it clear in documentation that `@reboot` means "run once after cron is started".

**Comment by divStar** [comment]
I would also like to know if there is a way to run a script only once. I do not exactly need a `@reboot` command or anything. The reason is: I use a container I created based on alpine and supercronic and on most occasions I use it to comfortably run scripts in a loop, but I've had occasions I only wanted to run the script once.

Currently I need to override the initial command in my `deployment.yaml`, but it'd be much nicer if I just could specify an environment variable or anything really to only execute a script once (a CLI command, that I could insert into my `query.sh` script would also be fine). I do not feel like using `pkill` or something like that.

**Comment by mmguero** [comment]
I'm still crossing my fingers on this. It's been a year since the last update, any movement on this (whether it's `@start` or `@reboot` or `@cronstart` I don't really care :D)

**Comment by github-cli** [comment]
and here I was migrating to supercronic and end up finding out that the reboot command does not work :-(

**Comment by github-cli** [comment]
seems like https://github.com/aptible/supercronic/pull/52#issue-458120854
never got finished :-(

**Comment by polarathene** [comment]
While probably not that helpful for `supercronic`, if you can run a process supervisor or init system on the container's entrypoint, you could then do something like this:

```ini
[program:my-recurring-command]
autorestart = true
stopasgroup = true
command = bash -c 'my-command && sleep 5m'
```

That's a snippet of the Python `supervisord` service config, paired with Bash to run a command at a recurring interval which works as follows:
- `my-command` executes.
- Wait 5 minutes and then exit the bash shell.
- `supervisord` restarts the `command` again, repeat.
- If any error occurs returning a non-zero exit status from `my-command`, it short circuits due to the `&&` and `supervisord` will try to restart it again (_it attempts 3 times before placing the service in a failed state_).
- If the service is stopped (`supervisorctl stop my-recurring-command`), it'll send a `SIGTERM` signal to the `bash` process and any other processes in that process group, such as `my-command` or `sleep` (whichever is actively processing), so no orphan concerns.
- **Bonus:**
  - No separate `@reboot` entry needed like cron requires to execute `my-command` straight-away (if you have other tasks queued with cron, you don't have to restart a whole daemon either to get that `@reboot` behaviour after container startup, no interruptions to other tasks_)
  - Intervals are reasonably consistent/deterministic, in the sense of once `my-command` completes, the interval is consistent.
    - With cron if your interval doesn't fit cleanly (eg: 25 minutes incremental), you get `my-command` run each hour at `:00` => `:25` => `:50` with the later only having 10 minutes between `:50` and the next `:00`.
    - Related to this is when a container starts you must wait N time until that scheduled interval is reached, or you have `@reboot` to get an immediate call, but depending on time of the container starting, you could have `my-command` called again shortly afterwards or some other offset that isn't your preferred interval since the last call of `my-command`.

Oddly there doesn't seem to be much options out there for a simple/portable CLI to run a command at a recurring interval, mostly just cron styled scheduling solutions. There are programs like `at` that require a daemon, or `ofelia` as a separate container, or systemd timers on the host, none of which I'd consider as convenient or appropriate for various reasons 😓 (_although system timers are quite nice, the lack of being able to bundle those into a container image that is not dependent upon container runtimes is a bummer_)

---

I thought that despite lack of `@reboot` handling, propagating `SIGTERM` might have been compatible with `supercronic` (_below an independent instance is called with an inlined crontab_):

```ini
[program:my-recurring-command]
autorestart = true
stopasgroup = true
# Bash is only needed here because of the `<( ... )` shell syntax:
command = bash -c 'exec supercronic --no-reap --quiet --passthrough-logs <(echo "*/5 * * * * my-command")'
```

However... [jobs are not run under the same process group](https://github.com/aptible/supercronic/blob/879f1ccf99dd03c252abe67f6f1265310a970025/cron/cron.go#L71-L73) for `supervisord` to signal nor does `supercronic` seem to forward `SIGTERM` to jobs (_which is presumably part of the [graceful shutdown approach](https://github.com/aptible/supercronic/blob/main/README.md#why-supercronic) while [waiting for active jobs to finish](https://github.com/aptible/supercronic/blob/879f1ccf99dd03c252abe67f6f1265310a970025/main.go#L269-L270) without giving them a heads-up?_), so if they take too long... a subsequent `SIGKILL` (_determined by `stopwaitsecs` setting in `supervisord`, default 10 secs_) terminates `supercronic` while the jobs it was managing continue to run orphaned until they exit. Thus `stopasgroup = true` in the supervisord service config has no effect, and be mindful of the `stopwaitsecs` value being sufficient for the jobs to complete in time.
